Unbind Cat pattern

devashishshankar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101767
* #101766
* #101765
* #101764

In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

devashishshankar committed:
Unbind Cat pattern

In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]

devashishshankar committed:
Update on "Unbind Cat pattern"


In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]

devashishshankar committed:
Update on "Unbind Cat pattern"


In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]

devashishshankar committed:
Update on "Unbind Cat pattern"


In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]

devashishshankar committed:
Update on "Unbind Cat pattern"


In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]

devashishshankar committed:
Update on "Unbind Cat pattern"


In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]

devashishshankar committed:
Update on "Unbind Cat pattern"


In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]

devashishshankar committed:
Update on "Unbind Cat pattern"


In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]

devashishshankar committed:
Update on "Unbind Cat pattern"


In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]

devashishshankar commented on torch/_inductor/fx_passes/pre_grad.py:
```diff
@@ -26,12 +26,14 @@
     merge_splits_pass,
     merge_split_cat_pass,
     split_squeeze_pass,
+    unbind_stack_pass,
```
So, I had a question on `PatternMatcherPass`. We iterate over graph.nodes while mutating the graph. Let's say I add a node during one of the mutations. Is it guaranteed that this node will be iterated over?

jansel replied:
No, if one pass depends on the changes made by another one they need to be seperate passes.

devashishshankar replied:
@jansel  So I see the following dependencies:

```
normalize_split_pass > merge_splits_pass > merge_split_cat_pass

normalize_split_pass > merge_splits_pass > split_squeeze_pass

split_squeeze_pass > unbind_stack_pass
```

I think possibly split_squeeze and unbind_stack can be combined into a single pass. But don't think others can be...

jansel replied:
Let's just have 4 passes not 5 then.

devashishshankar committed:
Update on "Unbind Cat pattern"


In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]

devashishshankar committed:
Update on "Unbind Cat pattern"


In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]

devashishshankar committed:
Update on "Unbind Cat pattern"


In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]

devashishshankar committed:
Update on "Unbind Cat pattern"


In continuation to previous diffs, this diff merges unbind->cat / unbind-> stack pattern.

In combination with previous diffs, this can handle split->squeeze->[cat/stack]

Since many of the cases are similar to split->cat, we reuse SplitCatSimplifier

Differential Revision: [D45955486](https://our.internmc.facebook.com/intern/diff/D45955486/)

[ghstack-poisoned]
